torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp

```diff
@@ -14,640 +16,1515 @@ namespace jit {
+
+  const int64_t warp_size =
+      n_elems * max_input_dtype_size * n_tensor_inputs < l2_cache_size
+      ? (int64_t)32 / max_input_dtype_size
+      : 32;
+
+  int64_t target_blocks = 1;
+  int64_t target_unroll = 1;
+  int64_t max_threads_in_block = warp_size;
+
+  // WARNING: Current device for codegen may not be the target device
+  const int64_t device_max_threads_per_multiprocessor =
+      (int64_t)at::cuda::getCurrentDeviceProperties()
+          ->maxThreadsPerMultiProcessor;
+
+  const int64_t device_multiprocessor_count =
+      (int64_t)at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
+
+  auto const max_unroll = ceilDiv(
+      // Available unrolling based on size of data type
+      (int64_t)16 / (int64_t)max_input_dtype_size,
+      // Reduce unrolling if we have many inputs, start reduction at 2 inputs
+      std::max((lastPow2((int64_t)n_tensor_inputs) >> 1), (int64_t)1));
+
+  // If we have one warp per block, how many blocks would that be?
+  target_blocks = ceilDiv(n_elems, (int64_t)warp_size);
+
+  // If we have more than a wave, put parallelism into unrolling
+  if (target_blocks > device_multiprocessor_count) {
+    target_unroll = std::min(
+        max_unroll, ceilDiv(target_blocks, device_multiprocessor_count));
+    target_blocks = ceilDiv(target_blocks, target_unroll);
+  }
+
+  // Cap target blocks to 4 waves
+  target_blocks = std::min(target_blocks, device_multiprocessor_count * 4);
+
+  if (target_blocks * target_unroll * max_threads_in_block < n_elems) {
+    // targetting 4 waves, so try to use a quarter of available threads
+    max_threads_in_block = std::min(
+        ceilDiv(n_elems, target_blocks * target_unroll),
+        ceilDiv(device_max_threads_per_multiprocessor, (int64_t)4));
+  }
+
+  // To get to target threads:
+  // Prioritize
+  // (1) x dim in iter domain
+  // (2) unrolling in iter domain
+  // (3) y in reduction domain
+  // To get target blocks:
+  // Prioritize
+  // (1) x dim in multiple outputs
+  // (2) y dim in multiple reductions - need to flip unrolling to reduction
+  // domain for this
+
+  // Blocks for outputs
+  int64_t gdimx = 1;
+
+  // Threads for reduction
+  int64_t bdimy = 1;
+  // Threads for output
+  int64_t bdimx = 1;
+
+  // Should we unroll from reduction axis, or outs axis
+  bool unroll_reduction = false;
+
+  bool multiple_reductions_per_block = true;
+
+  // Unroll amount
+  int64_t unroll_factor = 1;
+
+  int64_t remainder_in_reduction = num_elems_in_reduction;
+  int64_t remainder_in_output = num_outputs_for_reduction;
+
+  if (ceilDiv(num_outputs_for_reduction, warp_size) <
+      device_multiprocessor_count) {
+    // If we can't hit a full wave, reduce the warp_size to increase
+    // the number of blocks.  The warp should be reduced at a minimum
+    // to the granularity that an SM would pull a unique portion of a
+    // cacheline from the memory system or else there is no
+    // benefit from speading the work to a different block.
```

Comment:
Ugh, there is a minor typo in the comment.  `Speading` should be `spreading`.

Reply:
OOps, sorry I forgot to push my changes locally. I'll fix the typo in #851